examples/js/controls/OrbitControls.js

```diff
@@ -603,59 +683,157 @@ THREE.OrbitControls = function ( object, domElement ) {
 
-		var element = scope.domElement === document ? scope.domElement.body : scope.domElement;
+			case THREE.TOUCH_MODES.CLASSIC:
 
-		rotateLeft( 2 * Math.PI * rotateDelta.x / element.clientHeight ); // yes, height
+				rotateEnd.set( event.touches[ 0 ].pageX, event.touches[ 0 ].pageY );
 
-		rotateUp( 2 * Math.PI * rotateDelta.y / element.clientHeight );
+				rotateDelta.subVectors( rotateEnd, rotateStart ).multiplyScalar( scope.rotateSpeed );
 
-		rotateStart.copy( rotateEnd );
+				var element = scope.domElement === document ? scope.domElement.body : scope.domElement;
 
-		scope.update();
+				rotateLeft( 2 * Math.PI * rotateDelta.x / element.clientHeight ); // yes, height
 
-	}
+				rotateUp( 2 * Math.PI * rotateDelta.y / element.clientHeight );
 
-	function handleTouchMoveDollyPan( event ) {
+				rotateStart.copy( rotateEnd );
 
-		//console.log( 'handleTouchMoveDollyPan' );
+				break;
 
-		if ( scope.enableZoom ) {
+			case THREE.TOUCH_MODES.MAP:
 
-			var dx = event.touches[ 0 ].pageX - event.touches[ 1 ].pageX;
-			var dy = event.touches[ 0 ].pageY - event.touches[ 1 ].pageY;
+				// First finger
+				var x = event.touches[ 0 ].pageX;
+				var y = event.touches[ 0 ].pageY;
 
-			var distance = Math.sqrt( dx * dx + dy * dy );
+				rotateEnd.set( x, y );
+
+				// Second finger
+				var x2 = event.touches[ 1 ].pageX;
+				var y2 = event.touches[ 1 ].pageY;
+
+				rotateEnd2.set( x2, y2 );
+
+				// Angle between the two fingers at start and end
+				var angleStartFingers = rotateDelta.subVectors( rotateStart2, rotateStart ).angle();
+				var angleEndFingers = rotateDelta.subVectors( rotateEnd2, rotateEnd ).angle();
 
-			dollyEnd.set( 0, distance );
+				rotateDelta.subVectors( rotateEnd, rotateStart ).multiplyScalar( scope.rotateSpeed );
+				rotateDelta2.subVectors( rotateEnd2, rotateStart2 ).multiplyScalar( scope.rotateSpeed );
 
-			dollyDelta.set( 0, Math.pow( dollyEnd.y / dollyStart.y, scope.zoomSpeed ) );
+				// Angle between start and end position for each finger
+				var angleMoveFinger1 = rotateDelta.angle();
+				var angleMoveFinger2 = rotateDelta2.angle();
 
-			dollyIn( dollyDelta.y );
+				// We want to rotate up only if the two fingers move vertically without zooming or rotateLeft
+				// This end up by:
+				//  - angleStartFingers nearly colinear to x axis ==> fingers are aligned horizontally at start
+				//  - angleEndFingers nearly colinear to x axis ==> fingers are aligned horizontally at end
+				//  - angleMoveFinger1 nearly colinear to y axis ==> finger1 is moved vertically (ensure distance between finger is constant)
+				//  - angleMoveFinger2 nearly colinear to y axis ==> finger2 is moved vertically (ensure distance between finger is constant)
+				//  - Vector rotateStart --> rotateEnd and rotateStart2 --> rotateEnd2 are in the same way (prevent moving one finger vertically up while the other goes down)
 
-			dollyStart.copy( dollyEnd );
+				var epsilonAngle = 1 / 2; // ~= sin(PI / 6)
```

Comment:
This entire code block is too scary, IMO. Not sure what to do about it, though.

Reply:
Yeah, I'm not a big fan too

What about a refactor like this: 

```javascript
function handleTouchMoveRotate( event ) {

	//console.log( 'handleTouchMoveRotate' );

	if ( scope.enableRotate === false ) return;
	if ( ( state & STATE.ROTATE ) === 0 ) return;

	switch ( scope.touchMode ) {

		case THREE.TOUCH_MODES.CLASSIC:

			rotateEnd.set( event.touches[ 0 ].pageX, event.touches[ 0 ].pageY );

			rotateDelta.subVectors( rotateEnd, rotateStart ).multiplyScalar( scope.rotateSpeed );

			var element = scope.domElement === document ? scope.domElement.body : scope.domElement;

			rotateLeft( 2 * Math.PI * rotateDelta.x / element.clientHeight ); // yes, height

			rotateUp( 2 * Math.PI * rotateDelta.y / element.clientHeight );

			rotateStart.copy( rotateEnd );

			break;

		case THREE.TOUCH_MODES.MAP:

			// First finger
			var x = event.touches[ 0 ].pageX;
			var y = event.touches[ 0 ].pageY;

			rotateEnd.set( x, y );

			// Second finger
			var x2 = event.touches[ 1 ].pageX;
			var y2 = event.touches[ 1 ].pageY;

			rotateEnd2.set( x2, y2 );

			if ( isRotateUp() ) {

				var element = scope.domElement === document ? scope.domElement.body : scope.domElement;

				// rotating up and down along whole screen attempts to go 360, but limited to 180
				rotateUp( 2 * Math.PI * ( rotateEnd.y - rotateStart.y ) / element.clientHeight );

				// Start rotateUp ==> disable all movement to prevent flickering
				state = STATE.ROTATE_UP;

			} else if ( ( state & STATE.ROTATE_LEFT ) !== 0 ) {

				// Angle between the vector made by the two fingers and the positive x-axis at start
				var angleStartFingers = rotateDelta.subVectors( rotateStart2, rotateStart ).angle();
				// Angle between the vector made by the two fingers and the positive x-axis at end
				var angleEndFingers = rotateDelta.subVectors( rotateEnd2, rotateEnd ).angle();

				rotateLeft( ( angleStartFingers - angleEndFingers ) * scope.rotateSpeed );

			}

			rotateStart.copy( rotateEnd );
			rotateStart2.copy( rotateEnd2 );

			break;

		default:

			throw new Error( 'OrbitControls: Unexpected touchMode' );

	}

}

function isRotateUp() {

	// At start, does the two fingers are aligned horizontally
	if ( ! isHorizontal( rotateDelta.subVectors( rotateStart2, rotateStart ).angle(), Math.PI / 6 ) ) {

		return false;

	}

	// At end, does the two fingers are aligned horizontally
	if ( ! isHorizontal( rotateDelta.subVectors( rotateEnd2, rotateEnd ).angle(), Math.PI / 6 ) ) {

		return false;

	}

	// does the first finger moved vertically between start and end
	rotateDelta.subVectors( rotateEnd, rotateStart ).multiplyScalar( scope.rotateSpeed );
	if ( ! isVertical( rotateDelta.angle(), Math.PI / 6 ) ) {

		return false;

	}

	// does the second finger moved vertically between start and end
	rotateDelta2.subVectors( rotateEnd2, rotateStart2 ).multiplyScalar( scope.rotateSpeed );
	if ( ! isVertical( rotateDelta2.angle(), Math.PI / 6 ) ) {

		return false;

	}

	// Does the two finger moved in the same direction (prevent moving one finger vertically up while the other goes down)
	return rotateDelta.dot( rotateDelta2 ) > 0;

}

function isHorizontal( angleWithXAxis, precision ) {

	return Math.abs( Math.sin( angleWithXAxis ) ) < Math.sin( precision );

}

function isVertical( angleWithXAxis, precision ) {

	return Math.abs( Math.cos( angleWithXAxis ) ) < Math.cos( Math.PI / 2 - precision );

}
```